File: ios/ServerCase/Services/RemoteFiles.swift
import Foundation

struct RemoteFile: Identifiable, Hashable {
    var id: String { path }
    let name: String
    let path: String
    let isDirectory: Bool
    let isSymlink: Bool
    let sizeBytes: Int64
    let modifiedAt: Date
    /// Symbolic permission string, e.g. "rwxr-xr-x".
    let mode: String
}

struct RemoteListing {
    let path: String
    let entries: [RemoteFile]
}

enum RemoteFilesError: LocalizedError {
    case notADirectory(String)
    case command(String)

    var errorDescription: String? {
        switch self {
        case .notADirectory(let p): return "Cannot open \(p)"
        case .command(let m): return m
        }
    }
}

/// A remote file manager built on plain shell commands over the existing SSH
/// connection — the same "portable command + parse" approach the status
/// collector uses, so it needs nothing on the host beyond coreutils.
struct RemoteFiles {
    let service: SSHService

    /// Lists `path` (resolving it to an absolute path first).
    func list(_ path: String) async throws -> RemoteListing {
        let cmd = "cd \(quote(path)) 2>/dev/null && pwd && ls -lAL --time-style=+%s 2>/dev/null"
        let out = try await service.run(cmd)
        var lines = out.split(separator: "\n", omittingEmptySubsequences: false).map(String.init)
        guard let abs = lines.first, abs.hasPrefix("/") else {
            throw RemoteFilesError.notADirectory(path)
        }
        lines.removeFirst()
        var entries: [RemoteFile] = []
        for line in lines {
            if line.isEmpty || line.hasPrefix("total ") { continue }
            if let f = parse(line, in: abs) { entries.append(f) }
        }
        entries.sort {
            if $0.isDirectory != $1.isDirectory { return $0.isDirectory }
            return $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending
        }
        return RemoteListing(path: abs, entries: entries)
    }

    func readText(_ path: String) async throws -> String {
        try await service.run("cat \(quote(path))")
    }

    func writeText(_ path: String, _ content: String) async throws {
        let b64 = Data(content.utf8).base64EncodedString()
        _ = try await service.run("printf %s \(quote(b64)) | base64 -d > \(quote(path))")
    }

    func makeDirectory(_ path: String) async throws {
        _ = try await service.run("mkdir -p \(quote(path))")
    }

    func rename(_ from: String, to: String) async throws {
        _ = try await service.run("mv \(quote(from)) \(quote(to))")
    }

    func remove(_ path: String, isDirectory: Bool) async throws {
        _ = try await service.run((isDirectory ? "rm -r " : "rm -f ") + quote(path))
    }

    /// Downloads a remote file and writes it into the app's Documents dir,
    /// returning the local URL.
    func download(_ file: RemoteFile) async throws -> URL {
        let b64 = try await service.run("base64 \(quote(file.path))")
        guard let data = Data(base64Encoded: b64.replacingOccurrences(of: "\n", with: "")) else {
            throw RemoteFilesError.command("Failed to decode \(file.name)")
        }
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let url = docs.appendingPathComponent(file.name)
        try data.write(to: url)
        return url
    }

    /// Uploads raw bytes to `dir/name` on the remote host.
    func upload(_ data: Data, to dir: String, name: String) async throws {
        let b64 = data.base64EncodedString()
        _ = try await service.run("printf %s \(quote(b64)) | base64 -d > \(quote(join(dir, name)))")
    }

    func join(_ dir: String, _ name: String) -> String {
        let base = dir.hasSuffix("/") ? String(dir.dropLast()) : dir
        return base.isEmpty ? "/\(name)" : "\(base)/\(name)"
    }

    // MARK: Parsing

    private func parse(_ line: String, in dir: String) -> RemoteFile? {
        // perms links owner group size epoch name...
        let parts = line.split(separator: " ", omittingEmptySubsequences: true).map(String.init)
        guard parts.count >= 7 else { return nil }
        let perms = parts[0]
        let size = Int64(parts[4]) ?? 0
        let epoch = Double(parts[5]) ?? 0
        let name = parts[6...].joined(separator: " ")
        let isDir = perms.hasPrefix("d")
        let isLink = perms.hasPrefix("l")
        return RemoteFile(
            name: name,
            path: join(dir, name),
            isDirectory: isDir,
            isSymlink: isLink,
            sizeBytes: size,
            modifiedAt: Date(timeIntervalSince1970: epoch),
            mode: String(perms.dropFirst())
        )
    }

    private func quote(_ s: String) -> String {
        "'" + s.replacingOccurrences(of: "'", with: "'\\''") + "'"
    }
}

File: android/app/src/main/java/com/servercase/app/data/ssh/RemoteFiles.kt
package com.servercase.app.data.ssh

import java.nio.charset.StandardCharsets
import java.util.Base64

data class RemoteFile(
    val name: String,
    val path: String,
    val isDirectory: Boolean,
    val isSymlink: Boolean,
    val sizeBytes: Long,
    /** Modification time, epoch ms. */
    val modifiedAt: Long,
    /** Symbolic permission string, e.g. "rwxr-xr-x". */
    val mode: String,
)

data class RemoteListing(val path: String, val entries: List<RemoteFile>)

/**
 * A remote file manager built on plain shell commands over the existing SSH
 * connection — the same "portable command + parse" approach the status
 * collector uses, so it needs nothing on the host beyond coreutils.
 */
class RemoteFiles(private val client: SshClient) {

    suspend fun list(path: String): RemoteListing {
        val out = client.exec("cd ${quote(path)} 2>/dev/null && pwd && ls -lAL --time-style=+%s 2>/dev/null")
        val lines = out.split("\n")
        val abs = lines.firstOrNull()?.trim().orEmpty()
        require(abs.startsWith("/")) { "Cannot open $path" }
        val entries = lines.drop(1)
            .filter { it.isNotBlank() && !it.startsWith("total ") }
            .mapNotNull { parse(it, abs) }
            .sortedWith(compareByDescending<RemoteFile> { it.isDirectory }
                .thenBy(String.CASE_INSENSITIVE_ORDER) { it.name })
        return RemoteListing(abs, entries)
    }

    suspend fun readText(path: String): String = client.exec("cat ${quote(path)}")

    suspend fun writeText(path: String, content: String) {
        val b64 = Base64.getEncoder().encodeToString(content.toByteArray(StandardCharsets.UTF_8))
        client.exec("printf %s ${quote(b64)} | base64 -d > ${quote(path)}")
    }

    suspend fun makeDirectory(path: String) {
        client.exec("mkdir -p ${quote(path)}")
    }

    suspend fun rename(from: String, to: String) {
        client.exec("mv ${quote(from)} ${quote(to)}")
    }

    suspend fun remove(path: String, isDirectory: Boolean) {
        client.exec((if (isDirectory) "rm -r " else "rm -f ") + quote(path))
    }

    /** Downloads a remote file and returns its decoded bytes. */
    suspend fun download(file: RemoteFile): ByteArray {
        val b64 = client.exec("base64 ${quote(file.path)}").replace("\n", "")
        return Base64.getDecoder().decode(b64)
    }

    /** Uploads raw bytes to `dir/name` on the remote host. */
    suspend fun upload(bytes: ByteArray, dir: String, name: String) {
        val b64 = Base64.getEncoder().encodeToString(bytes)
        client.exec("printf %s ${quote(b64)} | base64 -d > ${quote(join(dir, name))}")
    }

    fun join(dir: String, name: String): String {
        val base = dir.trimEnd('/')
        return if (base.isEmpty()) "/$name" else "$base/$name"
    }

    private fun parse(line: String, dir: String): RemoteFile? {
        // perms links owner group size epoch name...
        val parts = line.trim().split(Regex("\\s+"))
        if (parts.size < 7) return null
        val perms = parts[0]
        val size = parts[4].toLongOrNull() ?: 0
        val epoch = parts[5].toLongOrNull() ?: 0
        val name = parts.drop(6).joinToString(" ")
        return RemoteFile(
            name = name,
            path = join(dir, name),
            isDirectory = perms.startsWith("d"),
            isSymlink = perms.startsWith("l"),
            sizeBytes = size,
            modifiedAt = epoch * 1000,
            mode = perms.drop(1),
        )
    }

    private fun quote(s: String): String = "'" + s.replace("'", "'\\''") + "'"
}
